Avalonia v11.0.0-preview4

DamianSuess opened the pull request:
Add support for Avalonia v11.0.0-preview4

## Action Items

* [ ] Add error catching and reporting for region registration
* [x] Test with [x:CompiledBindings="True"](https://docs.avaloniaui.net/docs/data-binding/compiledbindings)
    * This has been addressed by https://github.com/AvaloniaUI/Avalonia/pull/7246 
* [x] Test with HotReload NuGet, [Live.Avalonia](https://github.com/AvaloniaUI/Live.Avalonia)
* [x] Validate and update the Notification Service sample
    * Currently, Notifications are appearing in the top-left after the upgrade regardless of the position setting
* [x] Update ViewDiscovery sample which is not correctly rendering after the upgrade

## References
* https://github.com/AvaloniaUI/Avalonia/pull/8221
* https://github.com/AvaloniaUI/Avalonia/pull/8203
* https://github.com/AvaloniaUI/Avalonia/pull/9277

OmidID commented:
@DamianSuess Great job. Is it ready to merge?

DamianSuess commented:
Not yet, thanks for asking. Figured I'd make a PR marked as `draft` so you & others can see the progress of it and provide live feedback on any improvements.

DamianSuess commented:
Interesting error after upgrading to Avalonia v11.0.0-pre4...
Using the SampleMvvm project, the **DashboardView** causes an error to be thrown in the `RegionViewRegistry`'s `OnContentRegistered(..)` because the ListBox's `DataTemplate` does not define a `DataType` of the object in the template. Odd, we never had to do this before.. Even for a `List<string>`

Before:

```xaml
      <ListBox Margin="2"
               VerticalAlignment="Bottom"
               Items="{Binding ListItems}"
               ScrollViewer.HorizontalScrollBarVisibility="Visible"
               ScrollViewer.VerticalScrollBarVisibility="Visible"
               SelectedIndex="{Binding ListItemSelected}"
               SelectionMode="Single">
        <ListBox.DataTemplates>
          <DataTemplate>
            <TextBlock Text="{Binding .}"
                       FontSize="10"
                       TextWrapping="NoWrap" />
          </DataTemplate>
        </ListBox.DataTemplates>
```

DamianSuess commented:
@rabbitism, great find! I was digging thorough PRs & change docs trying to find this. Đông, you win the internet for the week 🚀 

I like the consistsncy, however, this new feature is going to be a surprise to some developers upgrading.

According to the PR's suggested error handling, it currently doesn't throw the exception during complie time. Only until `x:CompileBindings="true"` is set, otherwise we get it at runtime.

maxkatz6 commented on samples/ModulesSample/App.xaml.cs:
```diff
@@ -33,7 +33,7 @@ public static AppBuilder BuildAvaloniaApp()
                 .With(new Win32PlatformOptions
                 {
-                    EnableMultitouch = true,
+                    //// EnableMultitouch = true,  // Not supported in Avalonia v11.0.0-preview4
```
It is always enabled now

DamianSuess replied:
Thank you @maxkatz6 for the heads up. I updated the code comment to reflect your note

DamianSuess commented:
I fixe the ViewDiscovery sample to render, however, it is no longer auto-switching the views as it did w/ Avalonia v0.10.x. It appears that during the MainWindow's creation, it is not registering the `Prism Region ` "ContentRegion" in time.

To fix some of ViewDiscovery's AXML errors, I switched to Lite Theme, enabled compiled bindings, and moved View registration to the `App.xaml`. Still, the initial issue persists.

maxkatz6 commented on build/AvaloniaDependency.props:
```diff
@@ -3,14 +3,14 @@
-    <PackageReference Include="Avalonia.LinuxFramebuffer" Version="0.10.18" />
-    <PackageReference Include="Avalonia.Markup.Xaml.Loader" Version="0.10.18" />
-    <PackageReference Include="Avalonia.ReactiveUI" Version="0.10.18" />
+    <PackageReference Include="Avalonia" Version="11.0.0-preview4" />
+    <PackageReference Include="Avalonia.Controls.DataGrid" Version="11.0.0-preview4" />
+    <PackageReference Include="Avalonia.Desktop" Version="11.0.0-preview4" />
+    <PackageReference Include="Avalonia.Diagnostics" Version="11.0.0-preview4" />
+    <!--<PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.0.0-preview4" />-->
+    <PackageReference Include="Avalonia.LinuxFramebuffer" Version="11.0.0-preview4" />
+    <PackageReference Include="Avalonia.Markup.Xaml.Loader" Version="11.0.0-preview4" />
+    <PackageReference Include="Avalonia.ReactiveUI" Version="11.0.0-preview4" />
```
It seems to be the list of dependencies that will be referenced with the Prism.Avalonia nuget package.
If so, I believe this list is excessive.

Most likely, for the mvvm framework you most likely only need main "Avalonia" package.
Note, that I have used Prism a very little before in WPF, so I might miss some details.

- Avalonia.Desktop - will block application to run only on Win, Mac and Lin. With the way how .NET dependencies work, this dependency unfortunately won't allow users to target Browser (WASM) projects.
- Avalonia.Controls.DataGrid, Avalonia.Diagnostics and Avalonia.Markup.Xaml.Loader - these should be harmless in general, but also bloat application size, when they are not used (that's why  Diagnostics also disable for the release in our templates).
- Avalonia.LinuxFramebuffer - only makes sense if users want to target linux framebuffer, but should be harmless otherwise. 
- Avalonia.ReactiveUI - probably isn't needed here. If Prism uses System.Reactive, this dependency should be added manually (note, Avalonia doesn't reference System.Reactive anymore, starting with coming soon preview5)

maxkatz6 replied:
Some of these packages still are needed for the sample project though.

DamianSuess replied:
@maxkatz6 in the next preview 5 release of this repo, I'll trim those down. You're right, we should ***only*** be referencing the bare minimum, most of them aren't being used.

maxkatz6 commented on samples/SampleMvvmApp/Views/DashboardView.xaml:
```diff
@@ -46,13 +46,18 @@
-        <ListBox.DataTemplates>
-          <DataTemplate>
+        <ListBox.ItemTemplate>
+          <!--
+          OLD: <DataTemplate>
+          NEW: <DataTemplate DataType="{x:Type system:String}">
+          https://github.com/AvaloniaUI/Avalonia/pull/8221
+          -->
+            <DataTemplate DataType="{x:Type system:String}">
             <TextBlock Text="{Binding .}"
                        FontSize="10"
                        TextWrapping="NoWrap" />
           </DataTemplate>
-        </ListBox.DataTemplates>
+        </ListBox.ItemTemplate>
```
Note, you don't necessary need DataType, if you use ListBox.ItemTemplate instead ListBox.DataTemplates.
It won't cause any exceptions, as ListBox.ItemTemplate is way more specific and won't pollute children with the template.
Although, it's always a good idea to setup more strict types. Especially with x:CompileBindings.

maxkatz6 commented:
> I fixe the ViewDiscovery sample to render, however, it is no longer auto-switching the views as it did w/ Avalonia v0.10.x. It appears that during the MainWindow's creation, it is not registering the Prism Region  "ContentRegion" in time.

For this PR, can you try to switch to nightly builds? I remember there were issues with ContentControl which possibly can cause problems like this (note, I haven't debugged this issue).

Nightly builds, which should be shipped as preview5 in next weeks, some more breaking changes. But it also is expected to be last preview before release candidate.

More about nightly builds you can find here: https://github.com/AvaloniaUI/Avalonia/wiki/Using-nightly-build-feed or bot comments like this https://github.com/AvaloniaUI/Avalonia/pull/10058#issuecomment-1400691893

DamianSuess commented:
@maxkatz6, you're the man! i'll give it a shot. Wow, i wasn't expecting such quick feedback. If we ever meet up at some conference, a round of drinks are on me 👍

DamianSuess commented:
hello all, just a heads up on the slow down..
Though I prefer the keep work & personal life separate, I just want to let you all know for the past month I've been very ill and still recovering.  No, it's not the c-word.  Pneumonia hit really hard and took its toll.

When the spare time allows in between old man naps, I'll try and get out preview NuGet packages for you all to consume (_Preview-4, and [Nightly](https://github.com/AvaloniaUI/Avalonia/wiki/Using-nightly-build-feed)_) as per MaxKatz's awesome suggestion. I apologize for the inconvenience

DamianSuess commented:
Published v8.1.91.3-preview.11.4 to NuGet.org

* https://www.nuget.org/packages/Prism.Avalonia/8.1.97.3-preview.11.4
* https://www.nuget.org/packages/Prism.DryIoc.Avalonia/8.1.97.3-preview.11.4

OmidID commented:
WOA. WONDERFUL. THANKSSSSSSS A LOT. GREAT JOB
